### ISSUE TITLE
ld64.lld: error: too many personalities (4) for compact unwind to encode

### ISSUE THREAD

**Issue by glandium** [rust-lang/rust#102754]
When mixing ObjC, C++ and Rust in the same binary on aarch64-apple-darwin targets (e.g. like in Firefox), lld from LLVM 15 complains about "too many personalities". This is a regression from #92845 and affects beta.

For some reason, since #92845, this code in lld is hit: https://github.com/llvm/llvm-project/blob/d30dccd21f2c3bc1ed6cd054c131436a1af548e1/lld/MachO/UnwindInfoSection.cpp#L300

Ultimately, this leads to https://github.com/llvm/llvm-project/blob/d30dccd21f2c3bc1ed6cd054c131436a1af548e1/lld/MachO/UnwindInfoSection.cpp#L378
where there are 4 personalities instead of 3:
- `___objc_personality_v0`,
- `___gxx_personality_v0`,
- `_rust_eh_personality`
- `<internal>`

With previous rust versions, only the first three ones would be there. It looks like the relocation that triggers the `<internal>` is related to the `rust_eh_personality` function in `library/std/src/personality/gcc.rs` itself.

STR:
- Create a `bar.cc` file with content:
```
#include <stdexcept>

extern "C" void qux();

void bar() { qux(); throw std::runtime_error("foo"); }
```
- Create a `foo.m` file with content:
```
void foo() { @try {} @finally {} }
```
- Create a `qux.rs` file with content:
```
#![crate_type = "staticlib"]

extern "C" {
    fn foo();
}

#[no_mangle]
pub unsafe extern "C" fn qux() {
    println!("foo");
    foo();
}
```
- Create a `Makefile` file with content:
```
TARGET = --target=aarch64-apple-darwin

lib.dylib: bar.o foo.o libqux.a
	clang++ -o $@ $^ -fuse-ld=lld -dynamiclib $(TARGET) -lobjc

bar.o: bar.cc
	clang++ -o $@ -c $^ $(TARGET)

foo.o: foo.m
	clang -o $@ -c $^ $(TARGET)

libqux.a: qux.rs
	rustc $^ $(TARGET)
```
- Get clang 15 + lld and make sure it's in your `$PATH`.
- Set `SDKROOT` to point to a macOS SDK
- Run `make`

This does not happen on x86_64-apple-darwin, but interestingly the list of personalities there is:
- `___objc_personality_v0`,
- `___gxx_personality_v0`,
- `<internal>`
(no `_rust_eh_personality`).

With rust 1.64, the list of personalities was:
- `___objc_personality_v0`,
- `___gxx_personality_v0`,
- `_rust_eh_personality`

**Comment by nagisa** [comment]
So, uh, wouldn’t [this code](https://github.com/llvm/llvm-project/blob/d30dccd21f2c3bc1ed6cd054c131436a1af548e1/lld/MachO/UnwindInfoSection.cpp#L378) basically mean that you cannot link a binary that consists of more than 3 languages with their own unwinding personalities? That seems like an unfortunate limitation, and I would argue that this is something `ld.lld` should be fixing, regardless.

As for why the `<internal>` gets constructed, I would imagine that with the move of the personality into `libstd` one of the codegen units now starts containing both the personality function and the functionality that depends on that personality (whereas before nothing in the standalone unwinding crate could unwind and thus did not reference the personality function.)

A workaround on our end could probably be to make sure these personality functions end up in a CU of their own.

**Comment by nagisa** [comment]
Does this happen with other linkers (e.g. `ld` that ships with Xcode, if it is something other than `lld`?)

**Comment by glandium** [comment]
It doesn't happen with ld64, but the resulting binary only lists 3 personalities as per `llvm-object --unwind-info`...

**Comment by glandium** [comment]
Should this be labeled regression-stable-to-beta or something?

**Comment by glandium** [comment]
The LLVM bug has a fix for lld.

**Comment by bridiver** [comment]
> The LLVM bug has a fix for lld.

@glandium do you have a link for the bug fix?

**Comment by glandium** [comment]
https://reviews.llvm.org/D135728

**Comment by bridiver** [comment]
btw - why is `_rust_eh_personality` exported at all when panic == "abort"? It only applies to unwind, right?

**Comment by rillian** [comment]
> btw - why is `_rust_eh_personality` exported at all when panic == "abort"? It only applies to unwind, right?

[This comment](https://github.com/rust-lang/rust/blob/a54ddd4/library/panic_abort/src/lib.rs#L89-L114) says it's because the distributed core/std libraries are compiled with panic=unwind and need an `eh_personality` symbol to link to even if it's not going the used.

Now, maybe it shouldn't be exported from a staticlib with that resolution having already happened?

**Comment by bridiver** [comment]
```// To handle this the compiler only requires the `eh_personality` is defined if
// the panic runtime being linked in is the unwinding runtime, and otherwise
// it's not required to be defined (rightfully so). In this case, however, this
// library just defines this symbol so there's at least some personality
// somewhere.```
It seems like this is actually having the reverse of the intended effect. The real problem is the lld bug, but adding this in when it isn't actually required doesn't seem great to me.

**Comment by wesleywiser** [comment]
Does the lld fix need a backport to our LLVM fork?

**Comment by bjorn3** [comment]
`rust_eh_personality` will also be used in case of `-Cpanic=abort` for catching and aborting on exceptions crossing an `extern "C-unwind"` boundary in the future.

**Comment by bridiver** [comment]
@glandium did you test the llvm fix? It doesn't seem to be working for me for some reason. I just built using https://chromium.googlesource.com/a/external/github.com/llvm/llvm-project/+/refs/heads/llvmorg-16-init-8697-g60809cd2 which contains the fix https://chromium.googlesource.com/a/external/github.com/llvm/llvm-project/+/refs/heads/llvmorg-16-init-8697-g60809cd2/lld/test/MachO/eh-frame-personality-dedup.s
but I'm still getting the same error

**Comment by bridiver** [comment]
```
This reduces the number of distinct personalities we need in the final
binary, and helps us avoid hitting the "too many personalities" error.
```
this only handles the case of synthetic symbols

**Comment by glandium** [comment]
Yes, the fix works for us.

**Comment by bridiver** [comment]
> Yes, the fix works for us.

Strange, I tried with rustc 1.64.0 (a55dd71d5 2022-09-19) and clang version 16.0.0 (https://chromium.googlesource.com/a/external/github.com/llvm/llvm-project 60809cd29308e78d662375e6957a2b1b85415def) so I think there's still an issue here and we need a way to avoid including `_rust_eh_personality`. Including it by default causes problems that are far more difficult to fix than not including it.

**Comment by glandium** [comment]
1.64 didn't have the change that caused the problem for us in the first place.

**Comment by bridiver** [comment]
> 1.64 didn't have the change that caused the problem for us in the first place.

Everything after 1.59 was broken for us

**Comment by bridiver** [comment]
same thing with 1.65

**Comment by pnkfelix** [comment]
We've visited this during several T-compiler triage meetings, at this point I think the team agrees that this should be considered a P-high issue rather than a P-critical one. (So its still something we want to resolve, but we're not going to discuss it every week. And of course we're not blocking any releases over it.)

@rustbot label: -P-critical +P-high

**Comment by bridiver** [comment]
> We've visited this during several T-compiler triage meetings, at this point I think the team agrees that this should be considered a P-high issue rather than a P-critical one. (So its still something we want to resolve, but we're not going to discuss it every week. And of course we're not blocking any releases over it.)
> 
> @rustbot label: -P-critical +P-high

@pnkfelix any chance you would reconsider that? This just became a blocker for us for updating to chromium 109

**Comment by pnkfelix** [comment]
@bridiver I think it sounds like we need to dig into what specifically changed between 1.59 and 1.60 that broke things for you.

It sounds like the context of the failure you are observing differs from that described by this specific bug, since the workaround-fix did address things for @glandium 

do you have any kind of minimized reproduction, or a publicly linkable repository that I could use to observe your specific problem?

**Comment by bridiver** [comment]
@pnkfelix unfortunately I don't have a minimal repro, but cr110-rust branch of https://github.com/brave/brave-core fails with
```
ld64.lld: error: too many personalities (4) for compact unwind to encode
```
and cr110 does not. The only difference is the rust version (well, the cxx version is also different, but only because the current version of cxx doesn't build on 1.59)

**Comment by bridiver** [comment]
so the root of the problem is that _rust_eh_personality no longer gets stripped when using `-Clto` starting in 1.60 and so we still end up with too many in the final binary

**Comment by bridiver** [comment]
@pnkfelix I don't think it really makes sense in general to include https://github.com/rust-lang/rust/blob/a6269dad38e6ede0013ba3688099544833933c63/library/std/src/personality/emcc.rs#LL12C22-L12C41 for a static library, what it if was only included for executable targets?

**Comment by bjorn3** [comment]
The personality function is always needed if anything is compiled with `-Cpanic=unwind` (which the standard library is for most targets) as it is referenced by the unwind tables to support unwinding. Omitting it for static libraries will cause a linker error.

(By the way the specific function you are pointing too is only used for Emscripten. The personality function for most targets is at https://github.com/rust-lang/rust/blob/a6269dad38e6ede0013ba3688099544833933c63/library/std/src/personality/gcc.rs#LL244)

**Comment by bridiver** [comment]
> The personality function is always needed if anything is compiled with `-Cpanic=unwind` (which the standard library is for most targets) as it is referenced by the unwind tables to support unwinding. Omitting it for static libraries will cause a linker error.
> 
> (By the way the specific function you are pointing too is only used for Emscripten. The personality function for most targets is at https://github.com/rust-lang/rust/blob/a6269dad38e6ede0013ba3688099544833933c63/library/std/src/personality/gcc.rs#LL244)

Yes, I accidentally included the wrong link, but it's not necessary for static libraries because it can be included somewhere else in the final executable. This is exactly how it worked for us prior to 1.60 when it was stripped as unused code by lto

**Comment by bjorn3** [comment]
> Yes, I accidentally included the wrong link, but it's not necessary for static libraries because it can be included somewhere else in the final executable.

What would define the Rust personality function if not the Rust staticlib itself?

> This is exactly how it worked for us prior to 1.60 when it was stripped as unused code by lto.

What happened previously was that the unwind tables would not refer to the personality function. The linker doesn't complain about the existence of the `rust_eh_personality` function itself (which would be fixed by omitting the personality function), but the unwind tables refering to more than 3 personality functions in total (which won't be fixed by omitting the function itself, only by omitting the reference in the unwind table).

**Comment by bridiver** [comment]
> > Yes, I accidentally included the wrong link, but it's not necessary for static libraries because it can be included somewhere else in the final executable.
> 
> What would define the Rust personality function if not the Rust staticlib itself?
> 
> > This is exactly how it worked for us prior to 1.60 when it was stripped as unused code by lto.
> 
> What happened previously was that the unwind tables would not refer to the personality function. The linker doesn't complain about the existence of the `rust_eh_personality` function itself (which would be fixed by omitting the personality function), but the unwind tables refering to more than 3 personality functions in total (which won't be fixed by omitting the function itself, only by omitting the reference in the unwind table).

So what do you suggest to fix this problem?  Removing the one defined in chromium is not an option as that will cause other problems with stack traces.

**Comment by bjorn3** [comment]
I'm honestly don't know what the best solution is. Petitioning Apple to allow more than 3 personality functions in compact unwind info isn't very likely to succeed I presume. If you only need `-Cpanic=abort`, post-processing the rust staticlib to set the personality function in the unwind table to none in each object would be an option I guess at the cost of getting UB if a C++ exception reaches Rust code. Another option would be to try aliasing `rust_eh_personality` to `___gxx_personality_v0`. I believe the current implementation is a subset of `___gxx_personality_v0` and as such would work in the place of `rust_eh_personality`, but this may break in the future if we for example need some extra features `___gxx_personality_v0` doesn't provide. In addition I would like to see the `rust_eh_personality` symbol name to be versioned to prevent UB when linking libraries compiled by different rustc versions (for example a system library internally using rust compiled with an older rustc version and your own code) together.

**Comment by albertlarsan68** [comment]
> In addition I would like to see the `rust_eh_personality` symbol name to be versioned to prevent UB when linking libraries compiled by different rustc versions (for example a system library internally using rust compiled with an older rustc version and your own code) together.

There is already some mitigations in place (depending on the address of a static in std IIRC) to avoid this kind of UB

**Comment by bridiver** [comment]
In our case exceptions are not allowed in chromium and that personality routine exists to deal with non chromium code exceptions in a way that still produces a usable stack trace. Given that there are likely going to be other edge cases like this, would it not be possible to add an option that lets us handle it ourselves in the app that links to the static library? That seems like it would give us more flexibility to solve these kinds of problems.

**Comment by bjorn3** [comment]
> There is already some mitigations in place (depending on the address of a static in std IIRC) to avoid this kind of UB

That mitigation applies to one library catching a panic of another. It doesn't apply to the personality function being overriden by a different rustc version. (Overriding the personality function can only happen for staticlibs, not cdylibs)

**Comment by bjorn3** [comment]
> In our case exceptions are not allowed in chromium

That makes the problem a lot easier to solve. Does defining the personality function as an abort (+producing a stack trace depending on how exactly the crash handler is implemented) work? If so sharing the personality function between Rust and C++ shouldn't be an issue at all. It would just be the question how to make Rust libraries use this common aborting personality function. Rewriting the unwind tables should work, but is probably too slow. Maybe we could have a flag to specify the name of a personality function which is guaranteed to abort? This will require recompiling the standard library still as it has been compiled with references to rust_eh_personality. Or maybe someone else has a better idea? I would guess that omitting rust_eh_personality from the staticlib and defining it externally would still give this issue due to the symbol name being different from the other personality functions and thus producing a separate entry in the compact unwind info.

**Comment by bridiver** [comment]
> This will require recompiling the standard library still as it has been compiled with references to rust_eh_personality

or we could change the name of the personality function in chromium to rust_eh_personality

**Comment by bridiver** [comment]
> > This will require recompiling the standard library still as it has been compiled with references to rust_eh_personality
> 
> or we could change the name of the personality function in chromium to rust_eh_personality

In fact, if rust_eh_personality did the same thing as chromium's `CxxPersonalityRoutine` (https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame.cc?q=eh_frame&ss=chromium) for `panic=abort` then I think we could kill two birds with one stone. You would not have UB when mixing `panic=abort` and `panic=unwind` (library using panic=abort and std built with panic=unwind) and we could just use the rust personality routine instead of `CxxPersonalityRoutine`?

**Comment by bjorn3** [comment]
rust_eh_personality reads the same LSDA format as LLVM uses for C++ as this the only format LLVM supports. I believe it is a strict subset of what ___gxx_personality_v0 does, but I'm not entirely certain if it hasn't diverged in any way.

As for doing the same as CxxPersonalityRoutine, that would mean that the panic handler will no longer be invoked (in aborting mode) when unwinding into Rust in case of panic=abort, which means that the error message becomes less clear and there is no longer a backtrace when RUST_BACKTRACE=1 is set. I also don't think there is a way to distinguish `extern "C"` (which can't be unwound past) and `extern "C-unwind"` (which can be unwound past) in the future for `-Cpanic=unwind`, which would mean that we did have to present a different UX between the two.

Of these two options I think overriding rust_eh_personality to be what CxxPersonalityRoutine is now and using it in place of CxxPersonalityRoutine is the most realistic option. I think it did be nice if you could look for a symbol with a name containing rust_eh_personality, but necessary exactly equal. That way it would keep working if in the future a prefix or postfix unique for the rustc version is added to the symbol name. I don't think there is a need to change the rust_eh_personality part in the (near) future.

**Comment by bridiver** [comment]
If you look at what chromium did with the personality routine it's very interesting and maybe something similar could be implemented in rust so it can be easily changed at crate build time. See https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame_asm.S?q=eh_frame&ss=chromium

**Comment by bjorn3** [comment]
Rust handles the abort in the landingpad rather than personality functikn as LLVM doesn't provide a way to communicate to the personalty function if unwinding needs to abort or not. GCC does support this and if LLVM adds support for it, I think we might switch to it.

Chrome's method only works if unwinding will always abort, which isn't the case for rust in general. Without LLVM support for defining aborting "landingpads" it wouldn't be possible to do this in rustc without assembly though, even for `-Cpanic=abort`, because LLVM is allowed to omit the personality function specifier if there are no landingpads in the respective function as far as I know.

**Comment by bridiver** [comment]
Well, using llvm and assembly is exactly what chromium does, but I think it could work more easily if the personality routine was moved back out of std so it could be changed at crate build time. The problem now is that you have to unpack the archive and mess around with the symbols which is not ideal.

**Comment by bjorn3** [comment]
> but I think it could work more easily if the personality routine was moved back out of std so it could be changed at crate build time.

@Amanieu what do you think about this suggestion?

**Comment by Amanieu** [comment]
It seems to be like there are 2 reasonable solutions to @bridiver's problem:

1. Compile the Rust standard library with `-C panic=abort` using `-Z build-std`. However this is only possible on nightly Rust at the moment.

2. Fix Chrome to not use a custom personality function: the same effect as their custom personality function can be achieved with `__gxx_personality_v0` by emitting an empty call site table in the LSDA. This is what GCC emits for a `noexcept` function in C++ (but not Clang).

**Comment by bridiver** [comment]
> It seems to be like there are 2 reasonable solutions to @bridiver's problem:
> 
> 
> 
> 1. Compile the Rust standard library with `-C panic=abort` using `-Z build-std`. However this is only possible on nightly Rust at the moment.
> 
> 
> 
> 2. Fix Chrome to not use a custom personality function: the same effect as their custom personality function can be achieved with `__gxx_personality_v0` by emitting an empty call site table in the LSDA. This is what GCC emits for a `noexcept` function in C++ (but not Clang).

2 is not a viable option. I linked to the original issue where this change was discussed and it is necessary to get usable stack traces in certain situations. Also I believe that rust using the chromium personality routine (vs the other way around) may fix incorrect stack traces we sometimes get from rust in crash reporting.

**Comment by bridiver** [comment]
Also if I understand correctly in the future the personality function will also be included for panic=abort so even building std won't fix out problem. Effectively forcing this particular personality routine on anyone who wants to use rust doesn't seem great to me and in my opinion there should be a way to use your own. Having sensible defaults is one thing, but shouldn't the goal here be to make it easy for people to start integrating rust into existing code?

**Comment by Amanieu** [comment]
> 2 is not a viable option. I linked to the original issue [bugs.chromium.org/p/chromium/issues/detail?id=503128](https://bugs.chromium.org/p/chromium/issues/detail?id=503128) where this change was discussed and it is necessary to get usable stack traces in certain situations. Also I believe that rust using the chromium personality routine (vs the other way around) may fix incorrect stack traces we sometimes get from rust in crash reporting. We can very easily disable the personality function in chromium, but that will cause problems with crash reporting in some cases.

It absolutely is a viable option: if the call site table ([here](https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame_asm.S;l=98)) is empty, then you will hit [this case](https://github.com/llvm/llvm-project/blob/f42e1e0648779a55b7aff7e05a8a36c5a43b9a6a/libcxxabi/src/cxa_personality.cpp#L855) in `__gxx_personality_v0` which will force the terminate handler to be called. Notably, this will be called *before* stack unwinding takes place, so you will still have the full stack trace at the point the exception was thrown.

**Comment by bridiver** [comment]
> It absolutely is a viable option: if the call site table ([here](https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame_asm.S;l=98)) is empty, then you will hit [this case](https://github.com/llvm/llvm-project/blob/f42e1e0648779a55b7aff7e05a8a36c5a43b9a6a/libcxxabi/src/cxa_personality.cpp#L855) in `__gxx_personality_v0` which will force the terminate handler to be called. Notably, this will be called *before* stack unwinding takes place, so you will still have the full stack trace at the point the exception was thrown.

The issue discusses the problem with using the default handler. Without the personality routine that was added they were getting garbled stack traces in some instances. The problem is that there is some third party cpp code and objc used in chromium that throws exceptions, but chromium itself is built with noexceptions. Unless something has fundamentally changed about the default personality handler since this was implemented, we cannot remove it and still get valid stack traces in all instances.

**Comment by bridiver** [comment]
also what is the issue with adding some flexibility for cases you may not have anticipated? There are certainly personality routines used by applications that serve important purposes and can't just be removed to make room for the one rust provides. This limitation is unfortunate, but given that is probably isn't going away I think rust should accommodate the potential existence of personality routines in existing code.

**Comment by Amanieu** [comment]
I'm saying that you can get the same effect as chrome's custom personality function by using `__gxx_personality_v0` and modifying the assembly code a bit. This solves your problem which is that Apple's compact unwinding format only allows a maximum of 3 personality functions.

**Comment by bridiver** [comment]
> I'm saying that you can get the same effect as chrome's custom personality function by using `__gxx_personality_v0` and modifying the assembly code a bit. This solves your problem which is that Apple's compact unwinding format only allows a maximum of 3 personality functions.

Even if that's true for us, when the personality routine is also added for panic=abort mixed source apps will not be able to use a custom personality routine for any purpose. Doesn't that seem like a potential problem for rust that needs to be addressed independently of our specific issue?

**Comment by Amanieu** [comment]
I don't think this is a very common issue. In practice it is extremely rare for custom personality functions (outside of those used by LLVM-generated code) to be used. In fact this chrome code is the first time I have ever seen one. Also this issue only affects apple targets which further restricts the set of applications that could be affected by this.

**Comment by bridiver** [comment]
also we still have the issue where rust stack traces are sometimes garbled (which seems likely to be related) and I'm not sure how to address that with no way to change the rust personality routine. We can test it (for now) by building std with abort, but even that option will apparently go away soon and we would have to patch rust if we wanted to change it

**Comment by Amanieu** [comment]
I think you should open a separate issue for that.

**Comment by bridiver** [comment]
> I think you should open a separate issue for that.

I agree it's a separate issue, but I strongly suspect it's related.

**Comment by bridiver** [comment]
> I don't think this is a very common issue. In practice it is extremely rare for custom personality functions (outside of those used by LLVM-generated code) to be used. In fact this chrome code is the first time I have ever seen one. Also this issue only affects apple targets which further restricts the set of applications that could be affected by this.

With enough people using rust, uncommon situations will happen with increasing frequency. I'm really struggling to understanding why you'd want to keep this known limitation that some people will encounter. There are a lot of things I've never personally seen and yet I know they exist, sometimes in large numbers.

**Comment by bridiver** [comment]
> Also this issue only affects apple targets which further restricts the set of applications that could be affected by this.

I don't think that any cross platform app trivializes the set of applications that could be affected by this

**Comment by bridiver** [comment]
> I'm saying that you can get the same effect as chrome's custom personality function by using `__gxx_personality_v0` and modifying the assembly code a bit. This solves your problem which is that Apple's compact unwinding format only allows a maximum of 3 personality functions.

Can you explain a bit more here? I'm not seeing how modifying the assembly can produce the same effect without the custom personality function. There will be an exception handler which is the problem that the custom function addresses. From the issue:
```
When an exception occurs with CFRunLoop on the stack (as most of our code will have), then it will find the eh_frame from its @try/@catch block. The @catch(...)-all that CFRunLoopRunSpecific does then calls __cxa_rethrow, which begins again the proces of finding an exception handler. Since no other handler frame is on the stack, that will call std::terminate. But since this is a rethrow, the stack will only show up to where CFRunLoop caught the exception. This is what's producing our garbage stacks off the CFRunLoop. Basically, our code is no longer building as if -fno-exceptions were present, because we *do* enable ObjC exceptions (which are really the same thing).
```

```
This introduces base::mac::CallWithEHFrame(), which will run a block in a
stack frame that is set up to stop searching for an exception handler when it
is reached. This is used to prevent a top-level exception handler, installed
in CFRunLoopRunSpecific(), from catching and rethrowing C++ exceptions. When
it does this, the resulting stack traces are not useful for triage purposes. By
stoping the search for an exception handler, the runtime will call
std::terminate directly from the throwing stack trace.
```

**Comment by bjorn3** [comment]
As I understand if __gxx_personality_v0 is used with the call site table empty, the personality function will call `std::terminate` during the search phase when all stack frames still exist as the search phase is way before the personality function for `@catch` in `CFRunLoop` gets a change to run and catch the exception to rethrow it.

**Comment by bridiver** [comment]
> As I understand if __gxx_personality_v0 is used with the call site table empty, the personality function will call `std::terminate` during the search phase when all stack frames still exist as the search phase is way before the personality function for `@catch` in `CFRunLoop` gets a change to run and catch the exception to rethrow it.

but as it says in the issue there is an exception handler in CFRunLoopRunSpecific which causes the exception to be re-thrown and that's what causes the problem with the stack trace. My understanding is that the chromium personality function short circuits the exception handler search and then the default personality function works as you describe because it functions as if the call site table is empty when it's not.

**Comment by bridiver** [comment]
Could this problem maybe be solved by just marking `rust_eh_personality` as weak? Then it would be possible to override it in the final executable

**Comment by bridiver** [comment]
I think I found a hacky workaround by renaming `CxxPersonalityRoutine` -> `rust_eh_personality` and making it `extern __attribute__((weak)) "C"`. I also modified the corresponding asm and ensured that the static library containing the chromium definition was listed before the rust static library. I then threw an exception outside of a `CallWithEHFrame` block and got `*** Terminating app due to uncaught exception....` and inside the block I just got `Received signal 4 <unknown>...` so inside the block it's (correctly) not recognizing it as being caused by an exception.

**Comment by bridiver** [comment]
> I think I found a hacky workaround by renaming `CxxPersonalityRoutine` -> `rust_eh_personality` and making it `extern __attribute__((weak)) "C"`. I also modified the corresponding asm and ensured that the static library containing the chromium definition was listed before the rust static library. I then threw an exception outside of a `CallWithEHFrame` block and got `*** Terminating app due to uncaught exception....` and inside the block I just got `Received signal 4 <unknown>...` so inside the block it's (correctly) not recognizing it as being caused by an uncaught exception.
So actually this turns out not to work because it libbase.a (which contains the chromium function) and libbrave_rust.a (which is the static lib built by cargo) get linked into a framework and apparently that results in the same resolution (strong preferred) as linking to shared libraries that contain weak and strong symbols. When I linked libbase.a and libbrave_rust.a directly into a test exe everything worked as expected (first one listed was used), but no matter the order (I used a small test exe with only libbase.a and libbrave_rust.a), when linking to a shared lib that contains libbase.a and libbrave_rust.a, it always uses the strong symbol. It seems the only way we could make this work using this method is if rust exported the personality function as a weak symbol

**Comment by Amanieu** [comment]
@bridiver Have you tried my suggested solution?
- [Here](https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame_asm.S;l=98) delete everything between `Lcall_site_table_start` and `Lcall_site_table_end`.
- [Here](https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame_asm.S;l=23) change the personality to `___gxx_personality_v0`.

This should result in exactly the same behavior as the custom personality function while using one less total personality function in the program.

(Once this is confirmed to work, the assembly code can be further simplified, such as by removing the unnecessary catch clauses.)

**Comment by bridiver** [comment]
> @bridiver Have you tried my suggested solution?
> 
> * [Here](https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame_asm.S;l=98) delete everything between `Lcall_site_table_start` and `Lcall_site_table_end`.
> * [Here](https://source.chromium.org/chromium/chromium/src/+/main:base/mac/call_with_eh_frame_asm.S;l=23) change the personality to `___gxx_personality_v0`.
> 
> This should result in exactly the same behavior as the custom personality function while using one less total personality function in the program.
> 
> (Once this is confirmed to work, the assembly code can be further simplified, such as by removing the unnecessary catch clauses.)

That doesn't work, but I should have followed up because my description of the correct behavior was actually incorrect. I created a small test exe that contains only libbase.a (which contains the chromium personality function) and libbrave_rust.a and this is the output of triggering an objc exception inside `CallWithEHFrame` that is handled by the chromium personality function (libbase.a listed first). 

```
2023-01-27 07:39:37.333 a.out[55342:3250784] *** Terminating app due to uncaught exception 'FileNotFoundException', reason: 'File Not Found on System'
*** First throw call stack:
(
        0   CoreFoundation                      0x00007ff8136ab3eb __exceptionPreprocess + 242
        1   libobjc.A.dylib                     0x00007ff8131f7e25 objc_exception_throw + 48
        2   a.out                               0x000000010bdf7941 __main_block_invoke + 49
        3   a.out                               0x000000010bdf79c2 _ZN4base3mac15CallWithEHFrameEU13block_pointerFvvE + 10
        4   a.out                               0x000000010bdf7900 main + 16
        5   dyld                                0x00007ff813224310 start + 2432
)
libc++abi: terminating with uncaught exception of type NSException
```
With your proposed change what I get is 
```
Received signal 4 <unknown> 0001299fbe04
 [0x0001299ee122]
 [0x000129926023]
 [0x0001299ee071]
 [0x7ff813581c1d]
 [0x011000224900]
 [0x0001293fb63e]
 [0x000124e605f8]
 [0x000129394437]
 [0x00012939418f]
 [0x0001293929a2]
 [0x000129392bb0]
 [0x000124e5d66e]
 [0x00010bb5187e]
 [0x7ff813224310]
[end of stack trace]
[0131/090158.118930:WARNING:crash_report_exception_handler.cc(235)] UniversalExceptionRaise: (os/kern) failure (5)
zsh: illegal hardware instruction  
```
which is the same thing I am currently getting in Brave because there is still some dependency issue that is causing it to search libbrave_rust.a before libbase.a and also similar to what I get in my test exe if I list libbrave_rust.a before libbase.a

**Comment by bridiver** [comment]
I am not having any luck getting this to work. None of the above suggestions work and in general the inability to build a rust static library without rust_eh_personality is becoming really problematic for us.

**Comment by BrendanEich** [comment]
Cc'ing @pcwalton for help -- we use Rust in Brave's Chromium fork. IIUC, using C, Objective-C, and Rust in the same exe uses up all three personalities for compact unwind, and naively we need a fourth. Thanks.

**Comment by bridiver** [comment]
> The personality function is always needed if anything is compiled with `-Cpanic=unwind` (which the standard library is for most targets) or `extern "C-unwind"` is used[1](#user-content-fn-1-172db5c9b7bf636314f70de2559db4bf) as it is referenced by the unwind tables to support unwinding. Omitting it for static libraries will cause a linker error. In the past I believe LTO was allowed to strip out the personality function in case of `-Cpanic=abort`, but this is wrong when using `extern "C-unwind"` as we guarantee that unwinding into Rust code from `extern "C-unwind"` will abort when using `-Cpanic=abort`.
> 
> (By the way the specific function you are pointing too is only used for Emscripten. The personality function for most targets is at https://github.com/rust-lang/rust/blob/a6269dad38e6ede0013ba3688099544833933c63/library/std/src/personality/gcc.rs#LL244)
> 
> ## Footnotes
> 1. In the future it will even be used for `extern "C"` to ensure unwinding into rust code will abort when using `-Cpanic=abort`. This is a mitigation against UB. [↩](#user-content-fnref-1-172db5c9b7bf636314f70de2559db4bf)

@bjorn3 I tried build-std with panic=abort, but the "in the future" change is not in the future, that is exactly the change that broke things for us and rust_eh_personality is still exported https://github.com/rust-lang/rust/commit/5ff087669438ec406fd3abc99bd1b85b6fbe156a

**Comment by bridiver** [comment]
I found a workaround for now which is to use `build-std=panic_abort,std` (panic_abort may not be necessary) and patch gcc.rs in the downloaded source. It's not ideal, but it seems to be the least bad option atm.

**Comment by pcwalton** [comment]
That sounds like the right solution for now. I don't think that Rust aborts would have UB—they would kill the process safely.

(I don't have much to add to this issue beyond what's already been discussed, fwiw)

**Comment by bridiver** [comment]
> I don't think that Rust aborts would have UB—they would kill the process safely.

Are you sure about that? It sounds like the entire reason for making this change to move it into std and always include it is because it did have UB at least in some cases? In any case I got it working by leaving `rust_eh_personality` in place and doing a weak import on `rust_eh_personality_impl` which I defined in chromium to call `base::mac::CxxPersonalityRoutine`

**Comment by bridiver** [comment]
So based on what I have working, I think this could be very easily implemented by moving the implementation of `rust_eh_personality_impl` into a separate rlib that gets conditionally included based on some flag passed to cargo. If the rlib is excluded then it's up to the embedding app to provide an implementation or they'll get an undefined symbol

**Comment by XrXr** [comment]
I have a guess for a solution.  There is [`-C llvm-args=--emit-dwarf-unwind=always`](https://reviews.llvm.org/D122258) with newer LLVM 15 Rust. If you compile all Rust code with that option (`-Z build-std`, too) and then use `strip --remove-section` to yank all the compact unwind sections in the Rust output objects, I think that makes it so that `rust_eh_personality` is not a compact unwind personality. That should free up a spot.

The unwinder should be able to deal with DWARF unwind, because even when compact unwind is enabled, it can't encode all situations and needs DWARF as a fallback anyways.

**Comment by jyknight** [comment]
Ultimately, this is an Apple ABI issue, but I don't know whether that'll ever be fixed. They only allow 3 personalities to be used in compact-unwind in a single shared-object. And the system already has 3 canonical personalities already (`__gxx_personality_v0`, `__gcc_personality_v0`, and `__objc_personality_v0`). So, anyone using a custom personality is already doing something really fragile.

Because of this, I'd like to propose a change to LLVM to only emit compact unwind when using one of those 3 canonical Apple-shipped personality functions.

That should fix this issue "once and for all" (for non-rust uses as well), but does mean Rust code would use DWARF unwind for all functions -- even if the program being linked doesn't happen to require all 3 of the "official" personalities. Which is a bit unfortunate, but presumably not terrible.

**Comment by danakj** [comment]
This issue is now a blocker for Rust integration into Chromium mainline. We can try some of the workarounds here, and I am just starting to explore the space. Thanks for all the information here so far.

**Comment by oontvoo** [comment]
[The patch](https://reviews.llvm.org/rGe60b30d5e3878e7d91f8872ec4c4dca00d4a2dfc) implementing jyknight's proposal (with slight modification*) has landed. We should be able to have more than 3 personalities symbols now.

(*) In stead of making `__gxx_personality_v0`, `__gcc_personality_v0`, and `__objc_personality_v0` as "canonical", hence taking up all three compact-unwind encoding slots, the patch only considers `__gxx_personality_v0` and `__objc_personality_v0` to be canonical, leaving one "free" slot for custom personality symbol, if needed. The rationale for this is that `__gcc_personality_v0` is kind of rare compared to C++ and ObjC.

From a user's perspective, this feature requires no tweaking of any new flag - should "just work". 
However, if you have a custom personality function (eg., Rust's in this case) and you want it to  use compact-unwind, you can request that via the `-femit-compact-unwind-non-canonical` flag.

**Comment by StarMerely** [comment]
@oontvoo Hello, when we introduce the rust library into the iOS project, there will be a compilation error. I see that the above problem may be the same, but I tried it and it didn't work.

**Comment by apiraino** [comment]
Visiting again for triage.

Closing as per the previous [comment](https://github.com/rust-lang/rust/issues/102754#issuecomment-2827956244).
